### DISCUSSION TITLE
Is it possible for configuration changes to reflect in nixCats without rebuilding?

### DISCUSSION THREAD

**Discussion by meetmangukiya** [BirdeeHub/nixCats-nvim#28] [Q&A] [answered]
> The first main feature is the nixCats messaging system, which means you will not need to write ANY lua within your nix files (although you still can), and thus can use all the neovim tools like neodev that make configuring it so wonderful when configuring in your normal ~/.config/nvim

I initially thought it was possible for me to just tell nixCats the config dir in my config repo and it will create a wrapped nvim that will load lua code from the specified directory instead of the standard `~/.config/nvim`. But it looks like at the rebuild/switch it copies all the files in the `luaPath` over to the nix store and link that way, which means I have to build again for the config changes to reflect in the wrapped neovim? I naively tried to replace the [`cp -r ${luaPath}/* $out/`](https://github.com/BirdeeHub/nixCats-nvim/blob/main/nix/builder/default.nix#L109) with `ln -s ${luaPath} $out/` but that didn't really work(probably because nix wants to know definite hashes of the directories?).

In summary, is it possible to have live loading of lua code without having to rebuild nixcats if there is no nix change?

**Comment by BirdeeHub** [comment, marked as answer]
https://github.com/BirdeeHub/nixCats-nvim/blob/cd8e4193459a42b5cd6e21bc5c0f9155299a931e/flake.nix#L339

[:h nixCats.flake.outputs.settings](https://github.com/BirdeeHub/nixCats-nvim/blob/cd8e4193459a42b5cd6e21bc5c0f9155299a931e/nix/nixCatsHelp/nixCatsFlake.txt#L425-L508)

^ documentation

Yes, with wrapRc option. Obviously if you change any nix code you will still have to rebuild.

The quick tip in the help I linked is how I use the option. Basically, I just launch the version with `wrapRc = false` when im messing around with lua, and then rebuild when im done and go back to using the one with `wrapRc = true` so that it imports properly on other machines again. (The option just skips copying the directory and then you symlink it yourself)

Its really good for messing around with lua but because it doesnt copy it to the store it obviously wont be included if you do `nix run yourconfiglink` on another machine, hence why you would want 2 otherwise identical outputs, one with `wrapRc = true` and the other with `wrapRc = false`. Personally I have the `wrapRc = true` one aliased to vi and the `wrapRc = false` aliased to vim. That way I can use `vi .` normally, and when I want to mess with lua I can use `vim .` and see my changes reflected, and then rebuild and go back to using `vi .`

This has the added benefit that `vi .` is frozen how it was, meaning if I mess it up completely when messing with my lua in the `vim .` version I can open up `vi .` and fix it with a working neovim config rather than reverting all of what I just did just to be able to open the editor.

**Reply by meetmangukiya** [reply]
Thanks! Instead of manually symlinking my config repos lua directory to `~/.config/${configDirName}` we can edit nixCats(if we want to add such a feature to nixCats) to add luaPath to vim's `runtimepath` and in case of `wrapRc = false` just add a `require("${userRequire}")` in `runB4Config` so everything gets sourced as normal as if the config files were all in `~/.config/nvim` right?

**Reply by BirdeeHub** [reply]
If its not in the store, nix will not import it. You would have to symlink it either way. It would achieve the same thing but in more steps.

**Reply by BirdeeHub** [reply]
The home manager module could use mkOutOfStoreSymlink but it would apply only to the home manager module. Im not doing that because I feel thats confusing.

**Reply by BirdeeHub** [reply]
Actually I could maybe add another setting to the settings table that allows you to specify an absolute path instead of just setting NVIM_APPNAME? Maybe thats what you are asking? It still wouldnt import properly with `nix run`, but if you always clone your nix config to the same spot on every machine i guess like, maybe it could be useful. I'll think about it.

**Reply by BirdeeHub** [reply]
When I figure out what the heck to call a "absolute path when wrapRC is false" option, ill add it

Pretty sure its just 4 lines and 2-3 of those are just passing the option out to the spec XD

**Reply by BirdeeHub** [reply]
Any suggestions for a less verbose name for such an option? Im struggling

**Reply by BirdeeHub** [reply]
https://github.com/BirdeeHub/nixCats-nvim/blob/de6a409136575ec37f6f88b9841cd815c948ee2f/nix/builder/default.nix#L152-L156

Something like this?

**Reply by BirdeeHub** [reply]
The reason I did it like that is NVIM_APPNAME also sets the state directories and whatnot.

Unsure how to add this without it being a little weird. Im officially thinking about it though.

**Reply by BirdeeHub** [reply]
https://github.com/BirdeeHub/nixCats-nvim/blob/6958db3499c61673a79167bdafdfb6be4b7ce09a/nix/builder/default.nix#L150-L154

https://github.com/BirdeeHub/nixCats-nvim/blob/6958db3499c61673a79167bdafdfb6be4b7ce09a/nix/nixCatsHelp/nixCatsFlake.txt#L460-L464

I think this is better.

**Reply by BirdeeHub** [reply]
yeah Im ok with that. There ya go. Thats kinda the best I can do outside of mkOutOfStoreSymlink, which only works in home manager (and im not doing because different behaviour is generally bad behaviour)

**Reply by BirdeeHub** [reply]
I actually might use this option myself. Obviously I have no trouble symlinking it, but I clone my config to the same spot on every machine I own so... honestly may as well.

**Reply by BirdeeHub** [reply]
Keep in mind, it still wouldnt be runnable via nix run when wrapRc = false unless the config files were actually in the specified location. Thats what wrapRc is for. But on a machine you own, you will have your config cloned, and probably to the same spot every time so, may as well let you point at it.

**Reply by meetmangukiya** [reply]
I think this is working as I expected, thanks! Had to spend a bit of time debugging why the provided `unwrappedCfgPath` was not part of the `runtimepath`. Found lazy.nvim to be resetting it, need to provide `{ performance = { rtp = { reset = false } } }` in lazy's setup option. Although I know the point is to use nix for downloading the plugins instead of lazy being the package manager, I will probably do that incrementally. I am still figuring it out.

**Reply by BirdeeHub** [reply]
Yes, this is a known thing. The wrapper included in `nix flake init -t github:BirdeeHub/nixCats-nvim#luaUtils` actually does this for you.

https://github.com/BirdeeHub/nixCats-nvim/blob/646fba4e8c2aededf902c903a45cb3017f937916/nix/nixCatsHelp/luaUtils.txt#L59-L112

It sets this option for you, and does the reset it would have done but in a way that doesnt remove the config and treesitter grammars.


It also defines the plugin dev path in such a way that if you give it a list of plugin names to include (provided by `require('nixCats').pawsible.allPlugins.start` and `require('nixCats').pawsible.allPlugins.opt` and viewable with `:NixCatsPawsible`) it will look for them locally instead of downloading them, although preventing the rtp reset is the important part, this part can be skipped just fine if you choose to allow lazy to continue being the thing that downloads the plugins. (Using lazy to download them may occasionally cause failed build steps, which can be fixed either via adding missing dependencies to sharedLibraries and lspsAndRuntimeDeps sections, or downloading the thing that failed via nix and giving the wrapper its name)

You can see the implementation of that wrapper and its usage,

Usage can be seen here:

`nix flake init -t github:BirdeeHub/nixCats-nvim#kickstart-nvim`

https://github.com/BirdeeHub/nixCats-nvim/blob/646fba4e8c2aededf902c903a45cb3017f937916/nix/templates/kickstart-nvim/init.lua

(some of the things in that file are a little out of date, but everything about the lazy wrapper in it is correct, all that info is in the big banner comments, it will at least show you how to use the functions provided by the wrapper.)

(For lsps, there is a help page and a template for that too, and you can also take a look at the example in the main example config at root level of this repo at lua/myLuaConfig/LSPs. If you know how to use lspconfig it wont tell you anything new about lsps but its still helpful nonetheless)

Implementation here:

https://github.com/BirdeeHub/nixCats-nvim/blob/646fba4e8c2aededf902c903a45cb3017f937916/nix/templates/luaUtils/lua/nixCatsUtils/lazyCat.lua#L42-L136

As you can see it simply sets performance.rtp.reset = false, resets the rtp itself to include the nix provided paths still (other than the plugin path), and then sets dev.patterns and dev.path appropriately based on the list of names it is passed.

**Reply by BirdeeHub** [reply]
Also, thanks for inspiring the new option :) I can now include `unwrappedCfgPath = "/home/birdee/birdeeSystems/common/birdeevim";` instead of symlinking that directory to my `~/.config` directory. Its a small difference really but a good one.

Of course, this option is still only useful for fast iteration, as on another machine, if I did `nix run github:BirdeeHub/birdeeSystems#testvim` that file wouldnt be there and thus it would have no config, so I would want to run the one with `wrapRc = true`,  `nix run github:BirdeeHub/birdeeSystems#birdeeVim`instead.

**Reply by meetmangukiya** [reply]
For the runnability, maybe just dont commit `wrapRc = false; unwrappedCfgPath = ""` part and have that in a patch or stash locally. Not much difference really, just a workflow kind of thing maybe? Haven't really used [stacked git](https://stacked-git.github.io/) before, but I think it can be relevant for something like that.

**Reply by BirdeeHub** [reply]
I did that for a while but it was easier for me to just have 2 outputs, 1 with each, which include the same categories but with the 1 different setting

I factored it out so that I only have to change it 1 time but thats what nixCats and regularCats outputs in the example flake.nix is, just fully typed out

Heres how I factored it out (optional, you could just include the same categories in each):

https://github.com/BirdeeHub/birdeeSystems/blob/410a8f7959c3798b13c1944df245233cee8547ae/common/birdeevim/default.nix#L338C1-L379C5

I have categories and settings factored out like this

and then I just update with // in packageDefinitions

https://github.com/BirdeeHub/birdeeSystems/blob/410a8f7959c3798b13c1944df245233cee8547ae/common/birdeevim/default.nix#L406-L424

and then I have 2 outputs, 1 named birdeeVim and the other named testvim, one with wrapRc = true and one with wrapRc = false

I actually have both of them installed to my user, when I want the wrapped one I use vi and when I want the unwrapped one I use vim

That way I dont have to mess around with git

But yeah you can just output as many different versions of your config as you want which comes in handy for this :) mine has like 6 or 7 or something, some of which are installed some which arent, but all are exported from my system flake to be used on other machines.

**Reply by meetmangukiya** [reply]
Yeah, I will checkout how you are configuring it. Thanks for all the answers here!

**Reply by bartbie** [reply]
> Obviously if you change any nix code you will still have to rebuild.

YMMV but direnv should watch the nix files and rerun the shell on changes

**Reply by BirdeeHub** [reply]
you could yeah, it still takes a while to rebuild when its integrated into your whole system config so for best experience editing, not having to rebuild on lua changes is very nice.

**Reply by bartbie** [reply]
Oh yeah I don't disagree, I was thinking about using it along the wrapRC trick to get it both.

BTW, how can I change the config nixCats's nvim uses? I've tried to apply a hack I thought of recently - overriding neovim package with customRC option to source specific dir I want (in this case it's CWD since i use it for my dev shell) - however I couldn't make it work in nixCats. I tried providing custom unwrapped pkg like so but it doesn't seem to work lol

```nix
packageDefinitions.nvim-dev = {pkgs, ...}: {
  settings = { wrapRc = false; ...};
  neovim-unwrapped = pkgs.neovim.override {
   configure.customRC = ''
     source $SOMEVIMRC
   '';
  };
  ...
}
```

**Reply by BirdeeHub** [reply]
so, with wrapRc true, the path used is the path in the luaPath variable at the top of your template, which must be a store path (either a directory in your nix config, or from a flake input or something)

with wrapRc false it uses either ~/.config/nvim OR ~/.config/\<configDirName setting> OR unwrappedCfgPath depending on your settings in packageDefinitions for that package. see [:h nixCats.flake.outputs.settings](https://github.com/BirdeeHub/nixCats-nvim/blob/a41029f9d04f9f81eff45b2e74dab2f9ee0dc6b1/nix/nixCatsHelp/nixCatsFlake.txt#L501-L559) for all options and their default values. unwrappedCfgPath can be any path, although absolute paths are recommended obviously.

And remember, you can make as many packages as you want, making a second entry in packageDefinitions with the same categories and settings except wrapRc = false for experimenting with your lua is encouraged and very useful. However wrapRc = false will not import the config with the package because its not in the store, so you should have at least 1 package with wrapRc = true for running from anywhere via nix run

**Reply by BirdeeHub** [reply]
Also, about your "hack"

[You can import an already configured nixCats based config and RE configure parts of it,](https://github.com/BirdeeHub/nixCats-nvim/blob/a41029f9d04f9f81eff45b2e74dab2f9ee0dc6b1/nix/templates/REREREconfigure/flake.nix#L87-L100) and if you want to do that, you can still source new directories.

On the first configure you should use luaPath variable and the unwrappedCfgDir setting but when you RE configure it, you are using the original luaPath, so you may want to import a second directory for the extra config, and how to do that is demonstrated in that template, along with some other cool tricks.

Although that new second directory won't be affected by wrapRc setting and would always be loaded, so you wouldn't want to do it on the main configuration.

The template in the link isnt for beginner users, but for advanced users it shows some cool stuff you can do if you wanted to, say, import your nixcats config in a dev shell, make a modification, and then include it in the shell.

**Comment by BirdeeHub** [comment]
An aside because I think you also mentioned another thing 


If your Lua config is in a separate repo you can still put it in there you just have to import the repo. Just put it in flake inputs and put the input in the LuaPath variable. You could even do sub paths of said other repo with `"${inputs.theinput}/some/subpath"`. 

However this wouldn't solve the live update thing, because that particular version of said other repo is now in the store, so you would still need to use the wrapRc option or rebuild.

LuaPath option can only accept store paths, this is a hard limit of derivations. Derivations are built in a sandbox that doesn't have access to the rest of the filesystem (or the internet, although you can supposedly do some magic with fixed output derivations to get internet, you just have to specify the resulting hash of the final output of the derivation, and it will fail if it doesn't match. Pretty sure this is how they do some of the npm stuff. Normally the hash is of the derivation instructions but for fixed output ones, the hash is of the result.)
